### PR TITLE
[WIP] Behat refactor datagrid

### PR DESCRIPTION
**Pull request by SamirBoulil**
A proposal to refactor the behat datagrid and datagrid context using decorators.

### Definition

Let's introduce two definitions before starting:
- xyzCapableDecorators are used to decorate pages
- NodeElement decorators decorates low level node element

Capable decorators are classes which decorate an entire page and which responsibility is to provide access to (also decorated) nodeElement of interest within the page.
In addition to return this element, it will also decorate the nodeElement with the appropriated decorators for this element.

### Folder structure:

- Capable decorators should be placed in `features/Pim/Behat/Decorator/PageDecorator` folder.
- Element decorators are placed within a subfolder of `features/Pim/Behat/Decorator` depending on the context.
- Contexts are placed in `features/Pim/Behat/Context/Domain`:

## How do I use it ?

For instance, the `GridCapableDecorator` is capable to return the nodeElement representing the visible grid of the current page by calling `GridCapableDecorator::getCurrentGrid()`.
This method will return the nodeElement decorated by some grid compatible decorators like:
- `PaginationDecorator`: sets page size and page number of the grid
- `DataDecorator`: access the rows and columns of the grid
- `ActionDecorator`: provide access to massEdit actions for instance

In the context, I can access my GridCapableDecorator and my decorated node element grid using :
```php
$this
   ->getCurrentPage() // Returns a GridCapableDecorator instance
   ->getCurrentGrid() // Returns a nodeElement decorated with PaginationDecorator, DataDecorator...
   ->getPageSize() // Method of the PaginationDecorator
```

### Grid decorators
This PR introduces a page decorator to manipulate grids:
- [`Pim\Behat\Decorator\PageDecorator\GridCapableDecorator`](https://github.com/akeneo/pim-community-dev/blob/behat-refactor-datagrid/features/Pim/Behat/Decorator/PageDecorator/GridCapableDecorator.php): Capable of returning a decorated element representing the shown grid

This PR introduces the following grid decorators:
- [`Pim\Behat\Decorator\GridDecorator\DataDecorator`](https://github.com/akeneo/pim-community-dev/blob/behat-refactor-datagrid/features/Pim/Behat/Decorator/GridDecorator/DataDecorator.php): access the rows and columns of the grid
- [`Pim\Behat\Decorator\GridDecorator\ActionDecorator`](https://github.com/akeneo/pim-community-dev/blob/behat-refactor-datagrid/features/Pim/Behat/Decorator/GridDecorator/ActionDecorator.php): provide access to massEdit actions for instance
- [`Pim\Behat\Decorator\GridDecorator\FilterDecorator`](https://github.com/akeneo/pim-community-dev/blob/behat-refactor-datagrid/features/Pim/Behat/Decorator/GridDecorator/FilterDecorator.php): to manipulate filters (activate/deactivate as well as changing values)
- [`Pim\Behat\Decorator\GridDecorator\PaginationDecorator`](https://github.com/akeneo/pim-community-dev/blob/behat-refactor-datagrid/features/Pim/Behat/Decorator/GridDecorator/PaginationDecorator.php): to manipulate page size and page number

Basic usage: [`Pim/Behat/Context/Domain/Enrich/GridFilterContext.php#L30`](https://github.com/akeneo/pim-community-dev/blob/behat-refactor-datagrid/features/Pim/Behat/Context/Domain/Enrich/GridFilterContext.php#L30)
```php
    /**
     * @param string $filterName
     *
     * @Then /^I show the filter "([^"]*)"$/
     */
    public function iShowTheFilter($filterName)
    {
        $this->getCurrentPage()->getGridFilters()->showFilter($filterName);
    }
```

### View Decorators

This PR introduces a page decorator to manipulate views:
- [`Pim\Behat\Decorator\PageDecorator\ViewCapableDecorator`](https://github.com/akeneo/pim-community-dev/blob/behat-refactor-datagrid/features/Pim/Behat/Decorator/PageDecorator/ViewCapableDecorator.php): Capable of returning a decorated element representing the view selector

This PR introduces the following view decorator:
- [`Pim\Behat\Decorator\ViewDecorator\ViewSelectorDecorator`](https://github.com/akeneo/pim-community-dev/blob/behat-refactor-datagrid/features/Pim/Behat/Decorator/ViewDecorator/ViewSelectorDecorator.php): is capable to create, update, delete, find and apply a view with the view selector of the current page

Basic usage : [`features/Pim/Behat/Context/Domain/Enrich/ViewContext.php:28`](https://github.com/akeneo/pim-community-dev/blob/behat-refactor-datagrid/features/Pim/Behat/Context/Domain/Enrich/ViewContext.php#L28)
```php
    /**
     * @param string $viewLabel
     *
     * @When /^I apply the "([^"]*)" view$/
     */
    public function iApplyTheView($viewLabel)
    {
        $viewSelector = $this
               ->getCurrentPage() // Returns a ViewCapableDecorator
              ->getCurrentViewSelector(); // Returns a ViewSelectorDecorator
        $viewSelector->showViewList($viewLabel); // Calls shopViewList function of ViewSelectorDecorator
        $viewSelector->applyView($viewLabel);
        $this->wait();
    }
```

### The case of library managed html regions
Accessing node elements that are beyond the scope of a decorated low element can become an hassle. This can especially happen when you try to interact with html elements that are managed by an external library such as Select2.

Select2 uses javascript to manage an html list of values. In the dom, the list is appended to the Body tag in order for it to be well displayed and not dependant of the overall style of the page (CSS).

How do our low level decorated element supposed to access the select2 list when it is beyond it's scope (Meaning the nodeElement does not contain the list it needs to access) ?

The solution that is implemented here is to find the list by searching trough the parents of the node. For instance, to find the select2 list, we use an xpath expression to find the body tag and then within this body tag, find the list we want.

Example of implementation :
[`features/Pim/Behat/Decorator/GridDecorator/FiltersDecorator.php#L414`](https://github.com/akeneo/pim-community-dev/blob/behat-refactor-datagrid/features/Pim/Behat/Decorator/GridDecorator/FiltersDecorator.php#L414)
```php
    /**
     * Return the values proposed for the opened filter
     *
     * @return \Pim\Behat\Decorator\ElementDecorator
     */
    protected function getFilterAutocomplete()
    {
        return $this
            ->find('xpath', 'ancestor::body')
            ->find('css', $this->selectors['Filter list autocomplete']);
    }
````

I'd love to know what you think :) !

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | -
| Added Behats                      | Yes
| Changelog updated                 | -
| Review and 2 GTM                  | No
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

- [ ] Remove wait() calls
- [ ] Clean commits for code review
- [ ] Rebase master

### CODE REVIEW

**Comment by juliensnz** [comment]
This is a WIP so I didn't read the code in details but the described proposition is really nice. Not sure about the `Pim\Behat\Decorator\PageDecorator\FilterCapableDecorator` but I know that you have html constraint on this one. Maybe we can find a better solution.

File: features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    /**
+     * @When /^I press sequential-edit button$/
+     */
+    public function iPressSequentialEditButton()
+    {
+        $this->getCurrentPage()->sequentialEdit();
+        $this->wait();
+        $this->getNavigationContext()->currentPage = 'Product edit';
+    }
+}
```
Comment (juliensnz) [inline features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php]: please configure your IDE once for all to add a blank line at the end of a file

File: features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        $this->wait();
+    }
+
+    /**
+     * @When /^I press sequential-edit button$/
+     */
+    public function iPressSequentialEditButton()
+    {
+        $this->getCurrentPage()->sequentialEdit();
+        $this->wait();
```
Comment (juliensnz) [inline features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php]: we should remove that

File: features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        $this->getNavigationContext()->currentPage = 'Product edit';
```
Comment (juliensnz) [inline features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php]: not sure it should be in a generic context

File: features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        ];
+    }
+
+    /**
+     * @When /^I press mass-delete button$/
+     */
+    public function iPressMassDeleteButton()
+    {
+        $this->getCurrentPage()->massDelete();
+        $this->wait();
```
Comment (juliensnz) [inline features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php]: we should remove it

File: features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php (hunk 2 above)
Comment (SamirBoulil) [inline features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php]: Juste *Moved* code around as a first attempt to make the architecture work. Removing waits is in my next step :)

**Comment by ClementGautier** [comment]
Awsome PR.

I would renam some things imo:
* as page decorators are placed in a PageDecorator folder I would put node element decorator in a NodeElementDecorator folder.
* DataGenerator is too generic for me. as it basicly deals with cells of the grid I would rename it Grid or Table decorator.

That said I'm totaly :+1: for this PR.

File: features/Pim/Behat/Decorator/ViewDecorator/ViewSelectorDecorator.php [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+ * @author    Samir Boulil <samir.boulil@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ViewSelectorDecorator extends ElementDecorator
+{
+    use SpinCapableTrait;
+
+    protected $selectors = [
+        'View list' => '.ui-multiselect-menu.highlight-hover',
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/ViewDecorator/ViewSelectorDecorator.php]: the `.highlight-hover` seems weird. Are you sure ?

File: features/Pim/Behat/Decorator/ViewDecorator/ViewSelectorDecorator.php [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+     *
+     * @param $viewLabel
+     *
+     * @return NodeElement|null
+     *
+     * @throws \Context\Spin\TimeoutException
+     */
+    public function findView($viewLabel)
+    {
+        $viewList = $this
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/ViewDecorator/ViewSelectorDecorator.php]: yeah, this is the way we should go, even if it's not clean, it's due to the underlying lib (select2 in the case).

File: features/Pim/Behat/Decorator/ViewDecorator/ViewSelectorDecorator.php (hunk 5 above)
Comment (SamirBoulil) [inline features/Pim/Behat/Decorator/ViewDecorator/ViewSelectorDecorator.php]: This what we used here before :/
https://github.com/akeneo/pim-community-dev/blob/5ddd25d8cbdc70eaa0823380b71bcbe0700bab88/features/Context/Page/Base/Grid.php#L44

File: features/Pim/Behat/Context/Domain/Enrich/ViewContext.php [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+use Behat\Behat\Context\Step;
+use Behat\Behat\Context\Step\Then;
+use Behat\Gherkin\Node\TableNode;
+use Behat\Mink\Exception\ExpectationException;
+use Context\Spin\SpinCapableTrait;
+use Pim\Behat\Context\PimContext;
+
+/**
+ * A context for managing the grid pagination and size
```
Comment (grena) [inline features/Pim/Behat/Context/Domain/Enrich/ViewContext.php]: Wrong doc

File: features/Pim/Behat/Decorator/GridDecorator/ActionDecorator.php [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+ */
+class ActionDecorator extends ElementDecorator
+{
+    use SpinCapableTrait;
+
+    protected $selectors = [
+        'Grid'         => 'table.grid',
+        'Grid toolbar' => 'div.grid-toolbar',
+        'Mass Edit'    => '.mass-actions-panel .action i.icon-edit',
+        'Configure columns'     => ['css' => 'a:contains("Columns")']
```
Comment (grena) [inline features/Pim/Behat/Decorator/GridDecorator/ActionDecorator.php]: Does it work in French or German for example?

File: features/Pim/Behat/Decorator/ViewDecorator/ViewSelectorDecorator.php [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+            ->find('xpath', 'ancestor::body')
+            ->find('css', $this->selectors['View list']);
+
+        return $this->spin(function () use ($viewList, $viewLabel) {
+            return $viewList->find('css', sprintf('label:contains("%s")', $viewLabel));
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/ViewDecorator/ViewSelectorDecorator.php]: you should spin also on the view list

File: features/Pim/Behat/Decorator/ViewDecorator/ViewSelectorDecorator.php [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        $this->find('css', '#remove-view')->click();
+    }
+
+    /**
+     * Click on view in the view select
+     *
+     * @param string $viewLabel
+     *
+     * @throws \InvalidArgumentException
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/ViewDecorator/ViewSelectorDecorator.php]: you can drop it

File: features/Pim/Behat/Decorator/GridDecorator/ActionDecorator.php (hunk 8 above)
Comment (SamirBoulil) [inline features/Pim/Behat/Decorator/GridDecorator/ActionDecorator.php]: Idk just used the one we normally used :( : https://github.com/akeneo/pim-community-dev/blob/5ddd25d8cbdc70eaa0823380b71bcbe0700bab88/features/Context/Page/Base/Grid.php#L42

File: features/Pim/Behat/Decorator/PageDecorator/ViewCapableDecorator.php [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+use Pim\Behat\Decorator\ElementDecorator;
+
+/**
+ * Decorators of elements to manipulate page views
+ *
+ * @author    Samir Boulil <samir.boulil@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ViewCapableDecorator extends ElementDecorator
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/PageDecorator/ViewCapableDecorator.php]: GridViewCapableDecorator or at least in a Grid subnamespace

File: features/Pim/Behat/Decorator/PageDecorator/ViewCapableDecorator.php [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+    /**
+     * Finds and returns the element responsible for changing the view
+     *
+     * @return ElementDecorator
+     */
+    public function getCurrentViewSelector()
+    {
+        $element = $this->find('css', $this->selectors['View selector'])->getParent();
+        return $this->decorate($element, $this->decorators);
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/PageDecorator/ViewCapableDecorator.php]: always a blank line before a return

File: features/Pim/Behat/Decorator/PageDecorator/GridCapableDecorator.php [hunk 13]
```diff
@@ -46,6 +55,20 @@ function () {
...
+
+    /**
+     * Return the filters within the grid toolbar
+     *
+     * @return ElementDecorator
+     */
+    public function getGridFilters()
+    {
+        $filters = $this->getCurrentGrid()->find('css', $this->selectors['Filters']);
+        return $this->decorate($filters, $this->decorators['Filter decorators']);
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/PageDecorator/GridCapableDecorator.php]: same here, blank line before

File: features/Pim/Behat/Decorator/PageDecorator/GridCapableDecorator.php [hunk 14]
```diff
@@ -46,6 +55,20 @@ function () {
             'No visible grid found'
         );
 
-        return $this->decorate($grid->getParent()->getParent()->getParent(), $this->decorators);
+        return $this->decorate(
+            $grid->getParent()->getParent()->getParent()->getParent(),
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/PageDecorator/GridCapableDecorator.php]: any other solutions ? Like a getParents(".some_selector")

File: features/Pim/Behat/Decorator/PageDecorator/ViewCapableDecorator.php (hunk 11 above)
Comment (SamirBoulil) [inline features/Pim/Behat/Decorator/PageDecorator/ViewCapableDecorator.php]: I agree with you. 

The reason it is more generic here is that maybe *in a far far future* we can have a view switcher in a totally different page that does not have any grid on it.

The view switcher does not "care" about what view is shown, it just switches and apply views. 

Imvho, it better to keep it generic like this.

File: features/Pim/Behat/Decorator/PageDecorator/GridCapableDecorator.php [hunk 15]
```diff
@@ -20,17 +21,25 @@ class GridCapableDecorator extends ElementDecorator
...
 
     /** @var array */
     protected $decorators = [
-        'Pim\Behat\Decorator\GridDecorator\PaginationDecorator',
+        'Grid decorators'   => [
+            'Pim\Behat\Decorator\GridDecorator\DataDecorator',
+            'Pim\Behat\Decorator\GridDecorator\PaginationDecorator',
+            'Pim\Behat\Decorator\GridDecorator\ActionDecorator',
+        ],
+        'Filter decorators' => [
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/PageDecorator/GridCapableDecorator.php]: why do you split filterDecorators and GridDecorators ?

File: features/Pim/Behat/Decorator/GridDecorator/FilterDecorator.php [hunk 16]
```diff
@@ -0,0 +1,433 @@
...
+    }
+
+    /**
+     * Make sure a filter is visible
+     *
+     * @param string $filterName
+     *
+     * @throws \InvalidArgumentException
+     */
+    public function assertFilterNotVisible($filterName)
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/GridDecorator/FilterDecorator.php]: we should maybe put this back on context ? and only put the isVisible part here ?

File: features/Pim/Behat/Decorator/GridDecorator/FilterDecorator.php [hunk 17]
```diff
@@ -0,0 +1,433 @@
...
+    }
+
+    /**
+     * Make sure a filter is visible
+     *
+     * @param string $filterName
+     *
+     * @throws \InvalidArgumentException
+     */
+    public function assertFilterVisible($filterName)
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/GridDecorator/FilterDecorator.php]: same here. we could maybe put the method in the grid context and only implement the isVisible part here. We should never assert in decorators IMHO

File: features/Pim/Behat/Decorator/GridDecorator/FilterDecorator.php [hunk 18]
```diff
@@ -0,0 +1,433 @@
...
+            );
+        }
+    }
+
+    /**
+     * Make sure a filter is visible
+     *
+     * @param string $filterName
+     *
+     * @throws \InvalidArgumentException
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/GridDecorator/FilterDecorator.php]: you can remove this kind of crap

File: features/Pim/Behat/Decorator/GridDecorator/FilterDecorator.php [hunk 19]
```diff
@@ -0,0 +1,433 @@
...
+
+    /**
+     * @param string               $filterName The name of the filter
+     * @param string               $value      The value to filter by
+     * @param bool|string          $operator   If false, no operator will be selected
+     * @param DriverInterface|null $driver     Required to filter by multiple choices
+     *
+     * @throws \InvalidArgumentException
+     */
+    public function filterBy($filterName, $value, $operator = false, DriverInterface $driver = null)
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/GridDecorator/FilterDecorator.php]: Could be nice to rework this method but not the goal of this PR

File: features/Pim/Behat/Decorator/GridDecorator/DataDecorator.php [hunk 20]
```diff
@@ -0,0 +1,387 @@
...
+     * @param NodeElement $row
+     * @param int         $position
+     *
+     * @throws \InvalidArgumentException
+     *
+     * @return NodeElement
+     */
+    protected function getRowCell($row, $position)
+    {
+        // $row->findAll('css', 'td') will not work in the case of nested table (like proposals changes)
```
Comment (juliensnz) [inline features/Pim/Behat/Decorator/GridDecorator/DataDecorator.php]: you can remove that

File: features/Pim/Behat/Decorator/GridDecorator/FilterDecorator.php (hunk 19 above)
Comment (SamirBoulil) [inline features/Pim/Behat/Decorator/GridDecorator/FilterDecorator.php]: I agree

File: features/Pim/Behat/Decorator/PageDecorator/GridCapableDecorator.php (hunk 14 above)
Comment (SamirBoulil) [inline features/Pim/Behat/Decorator/PageDecorator/GridCapableDecorator.php]: Yes I wanted to rework this line.
unforunately: getParent() does not accept a locator : http://apigen.juzna.cz/doc/Behat/Mink/source-class-Behat.Mink.Element.NodeElement.html#48-56
In master we use this right now. 

The best alternative I find is :
- `$grid->find('xpath', '../../../../')`
- `$grid->find('xpath' 'ancestor::contains(@class, "className")')

File: features/Pim/Behat/Context/NavigationContext.php [hunk 21]
```diff
@@ -64,6 +64,7 @@ class NavigationContext extends PimContext implements PageObjectAwareInterface
     /** @var array */
     protected $pageDecorators = [
         'Pim\Behat\Decorator\PageDecorator\GridCapableDecorator',
+        'Pim\Behat\Decorator\PageDecorator\ViewCapableDecorator',
```
Comment (juliensnz) [inline features/Pim/Behat/Context/NavigationContext.php]: GridViewCapableDecorator or in a sub grid namespace :)

File: features/Pim/Behat/Context/Domain/Enrich/ViewContext.php [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+ *
+ * @author    Samir Boulil <samir.boulil@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ViewContext extends PimContext
```
Comment (juliensnz) [inline features/Pim/Behat/Context/Domain/Enrich/ViewContext.php]: same here

File: features/Pim/Behat/Context/Domain/Enrich/GridFilterContext.php [hunk 23]
```diff
@@ -0,0 +1,276 @@
...
+use Pim\Behat\Context\PimContext;
+
+/**
+ * A context for managing the grid pagination and size
+ *
+ * @author    Samir Boulil <samir.boulil@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class GridFilterContext extends PimContext
```
Comment (juliensnz) [inline features/Pim/Behat/Context/Domain/Enrich/GridFilterContext.php]: this class should be rework to remove a lot of code but it's not the goal of this PR so it's ok

File: features/Pim/Behat/Context/Domain/Enrich/GridDataContext.php [hunk 24]
```diff
@@ -0,0 +1,429 @@
...
+     * @throws ExpectationException
+     *
+     * @When /^I uncheck the rows? "([^"]+)"$/
+     */
+    public function iUncheckTheRows($rows)
+    {
+        $rows = $this->getMainContext()->listToArray($rows);
+
+        foreach ($rows as $row) {
+            $gridRow  = $this->getCurrentPage()->getCurrentGrid()->getRow($row);
```
Comment (juliensnz) [inline features/Pim/Behat/Context/Domain/Enrich/GridDataContext.php]: you have clean method in decorator now ?

File: features/Pim/Behat/Context/Domain/Enrich/GridFilterContext.php (hunk 23 above)
Comment (SamirBoulil) [inline features/Pim/Behat/Context/Domain/Enrich/GridFilterContext.php]: Agree, I thought about doing a FilterManagerDecorator and FilterDecorator to separate the activation/deactivation of filters from the filtering process itself. What do you think @juliensnz

File: features/Pim/Behat/Context/NavigationContext.php (hunk 21 above)
Comment (SamirBoulil) [inline features/Pim/Behat/Context/NavigationContext.php]: As discussed, the "ViewCapableDecorator" will stay be a generic page decorator but renamed as "UserViewConfigurationCapableDecorator" as it is clearer.
@juliensnz

File: features/Pim/Behat/Context/Domain/Enrich/ViewContext.php (hunk 22 above)
Comment (SamirBoulil) [inline features/Pim/Behat/Context/Domain/Enrich/ViewContext.php]: Agreed, will be renamed UserViewConfigurationContext

File: features/Pim/Behat/Context/Domain/Enrich/GridDataContext.php (hunk 24 above)
Comment (SamirBoulil) [inline features/Pim/Behat/Context/Domain/Enrich/GridDataContext.php]: Right, let's uncheck them all with ->check($row, false);

File: features/Pim/Behat/Decorator/ViewDecorator/ViewSelectorDecorator.php (hunk 9 above)
Comment (SamirBoulil) [inline features/Pim/Behat/Decorator/ViewDecorator/ViewSelectorDecorator.php]: Ok

File: features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php [hunk 25]
```diff
@@ -0,0 +1,210 @@
...
+    {
+        $this->getCurrentPage()->getCurrentGrid()->clickOnResetButton();
+    }
+
+    /**
+     * @Then /^I refresh the grid$/
+     */
+    public function iRefreshTheGrid()
+    {
+        $this->getCurrentPage()->getCurrentGrid()->clickOnRefreshButton();
```
Comment (juliensnz) [inline features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php]: getRefreshButton()->click()

File: features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php [hunk 26]
```diff
@@ -0,0 +1,210 @@
...
+            );
+        }
+    }
+
+    /**
+     * @Then /^I reset the grid$/
+     */
+    public function iResetTheGrid()
+    {
+        $this->getCurrentPage()->getCurrentGrid()->clickOnResetButton();
```
Comment (juliensnz) [inline features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php]: same here

File: features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php [hunk 27]
```diff
@@ -0,0 +1,210 @@
...
+            new Step\Then('I press mass-edit button')
+        ];
+    }
+
+    /**
+     * @When /^I press mass-edit button$/
+     */
+    public function iPressMassEditButton()
+    {
+        $this->getCurrentPage()->getCurrentGrid()->massEdit();
```
Comment (juliensnz) [inline features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php]: getBulkAction('mass-edit')->click() or start

File: features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php [hunk 28]
```diff
@@ -0,0 +1,210 @@
...
+
+    /**
+     * @param string $entities
+     *
+     * @Then /^I select rows? (.*)$/
+     */
+    public function iSelectRows($entities)
+    {
+        foreach ($this->getMainContext()->listToArray($entities) as $entity) {
+            $this->getCurrentPage()->getCurrentGrid()->selectRow($entity, true);
```
Comment (juliensnz) [inline features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php]: the boolean sounds weird

File: features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php [hunk 29]
```diff
@@ -0,0 +1,210 @@
...
+            new Step\Then('I press mass-delete button')
+        ];
+    }
+
+    /**
+     * @When /^I press mass-delete button$/
+     */
+    public function iPressMassDeleteButton()
+    {
+        $this->getCurrentPage()->getCurrentGrid()->massDelete();
```
Comment (juliensnz) [inline features/Pim/Behat/Context/Domain/Enrich/GridActionContext.php]: getBulkAction()->click()

File: features/Pim/Behat/Context/Domain/Enrich/GridDataContext.php (hunk 24 above)
Comment (juliensnz) [inline features/Pim/Behat/Context/Domain/Enrich/GridDataContext.php]: nice idea

File: features/Pim/Behat/Context/Domain/Enrich/GridFilterContext.php (hunk 23 above)
Comment (juliensnz) [inline features/Pim/Behat/Context/Domain/Enrich/GridFilterContext.php]: good idea. We could also hide this by doing something like `->getFilter('description')->filter('=', 'toto')`
`getFilter()` could get the filter (activate it if needed) and decorate it with the type of data of the attribute.

File: features/Pim/Behat/Context/Domain/Enrich/GridFilterContext.php [hunk 30]
```diff
@@ -0,0 +1,275 @@
...
+     *
+     * @throws ExpectationException
+     *
+     * @Then /^I should( not)? see the filters? (.*)$/
+     */
+    public function iShouldSeeTheFilters($not, $filters)
+    {
+        $toBeSeen = !(bool)$not;
+        $filters = $this->getMainContext()->listToArray($filters);
+        $gridFilters = $this->getCurrentPage()->getCurrentGrid();
```
Comment (juliensnz) [inline features/Pim/Behat/Context/Domain/Enrich/GridFilterContext.php]: $grid

File: features/Pim/Behat/Context/Domain/Enrich/GridFilterContext.php [hunk 31]
```diff
@@ -0,0 +1,275 @@
...
+
+        foreach ($filters as $filter) {
+            if ($toBeSeen) {
+                $gridFilters->assertFilterVisible($filter);
```
Comment (juliensnz) [inline features/Pim/Behat/Context/Domain/Enrich/GridFilterContext.php]: not sure the decorator should implement an assert method

File: features/Pim/Behat/Context/Domain/Enrich/GridFilterContext.php [hunk 32]
```diff
@@ -0,0 +1,275 @@
...
+     *
+     * @throws ExpectationException
+     *
+     * @Given /^I should( not)? see the available filters (.*)$/
+     */
+    public function iShouldSeeTheAvailableFilters($not, $filters)
+    {
+        $available = !(bool)$not;
+        $filters = $this->getMainContext()->listToArray($filters);
+        $gridFilters = $this->getCurrentPage()->getCurrentGrid();
```
Comment (juliensnz) [inline features/Pim/Behat/Context/Domain/Enrich/GridFilterContext.php]: $grid

**Comment by a2xchip** [comment]
@SamirBoulil I like this refactoring. Is there other way or ideas with select filters, as xpath usage doesn't seems so clear for me? Or may be it was just best usage example.

**Comment by SamirBoulil** [comment]
@a2xchip Thanks ! I appreciate that :)

This is the best solution I could come up with to access select2 elements (which is located as a direct children of the body). 
Even If I fill I'm bending the decorator system here, I can live with this xpath request *as long as* it is justified because it could also be a sign of a wrong page decorator design (ie,the page is not cut the correct way for what we want to do).

I hope I'll be able to merge real soon.

**Comment by a2xchip** [comment]
@SamirBoulil what is going on with this?

**Comment by SamirBoulil** [comment]
Hello Alexander,

It's just this PR is very old now and parallel efforts has been made since to stabilise the behats architecture.
The code present in this PR is outdated, I just close the PR to clean up a bit.

**Comment by a2xchip** [comment]
@SamirBoulil Thx for information. Have not seen behats state in 1.6 yet. Hope previous problems went away :+1:

**Pull request closed without merging**